File: models/Gpt-3.5-turbo/cpp/code/problems/dynamic_programming/Counting_Tilings.cpp

#include <iostream>
#include <vector>

using namespace std;

const int MOD = 1e9 + 7;

int countWays(int n, int m) {
    vector<vector<int>> dp(n + 1, vector<int>(m + 1));

    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= m; j++) {
            if (i == 1) {
                dp[i][j] = 1;
            } else if (j == 1) {
                dp[i][j] = 1;
            } else {
                dp[i][j] = (dp[i - 1][j] + dp[i][j - 2]) % MOD;
            }
        }
    }

    return dp[n][m];
}

int main() {
    int n, m;
    cin >> n >> m;

    cout << countWays(n, m) << endl;

    return 0;
}

File: models/Gpt-3.5-turbo/py/code/dynamic_programming/Counting_Tilings.py

MOD = int(1e9 + 7)

def count_ways(n, m):
    dp = [[0] * (m + 1) for _ in range(n + 1)]

    # Base cases
    dp[1][1] = 1
    dp[1][2] = 1
    dp[2][1] = 1
    dp[2][2] = 2

    for i in range(1, n + 1):
        for j in range(1, m + 1):
            if i != 1 or j != 1:  # Skip base cases
                dp[i][j] = (dp[i-1][j] + dp[i][j-1]) % MOD

    return dp[n][m]

# Read input
n, m = map(int, input().split())

# Calculate and print the number of ways
print(count_ways(n, m))
